apiserver/cloud/instance_information.go

```diff
@@ -0,0 +1,67 @@
+// Copyright 2016 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package cloud
+
+import (
+	"github.com/juju/errors"
+	"github.com/juju/juju/apiserver/common"
+	"github.com/juju/juju/apiserver/params"
+	"github.com/juju/juju/constraints"
+	"github.com/juju/juju/environs"
+	"github.com/juju/juju/state"
+	"github.com/juju/juju/state/stateenvirons"
+	names "gopkg.in/juju/names.v2"
+)
+
+// EnvironConfigGetter implements environs.EnvironConfigGetter
+// in terms of a *state.State.
+type cloudEnvironConfigGetter struct {
+	*state.State
```

Comment:
You only need state.CloudAccessor and the GetModel function. Backend (apiserver/cloud/backend.go) already embeds state.CloudAccessor, so you just need to add the GetModel method, and you can use Backend instead of State.

You could otherwise just ignore the ModelTag, and use the existing Backend.ControllerModel method.

I think it might be simpler though if you just pass the CloudSpec (or perhaps Environ?) into the common.InstanceTypes function? i.e. have the facades pull apart the args, get the appropriate model, cloud, and region, and construct a CloudSpec from them. You may as well then get the Environ then too, and avoid passing the backend in just to get ModelConfig?